// sanity-plugin-note-field/dist/note.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _ui = require("@sanity/ui");

var NoteField = /*#__PURE__*/_react["default"].forwardRef(function (_ref, ref) {
  var type = _ref.type;
  var options = type.options;
  var icon = options.icon,
      headline = options.headline,
      message = options.message,
      tone = options.tone;
  if (!message) return null;
  var CustomIcon = icon;
  return /*#__PURE__*/_react["default"].createElement(_ui.Card, {
    ref: ref,
    padding: [3, 3, 4],
    radius: 3,
    shadow: 1,
    tone: tone || 'primary'
  }, headline && /*#__PURE__*/_react["default"].createElement(_ui.Box, {
    marginBottom: 3
  }, /*#__PURE__*/_react["default"].createElement(_ui.Inline, {
    space: [1]
  }, icon && /*#__PURE__*/_react["default"].createElement(CustomIcon, {
    style: {
      fontSize: 24
    }
  }), /*#__PURE__*/_react["default"].createElement(_ui.Heading, {
    size: 1
  }, headline))), /*#__PURE__*/_react["default"].createElement(_ui.Flex, null, /*#__PURE__*/_react["default"].createElement(_ui.Box, {
    style: {
      flexShrink: 0
    }
  }, icon && !headline && /*#__PURE__*/_react["default"].createElement(CustomIcon, {
    style: {
      fontSize: 24
    }
  })), /*#__PURE__*/_react["default"].createElement(_ui.Box, {
    marginLeft: !headline ? 3 : 0
  }, /*#__PURE__*/_react["default"].createElement(_ui.Text, {
    size: [1, 1, 1]
  }, message))));
});

var _default = NoteField;
exports["default"] = _default;
