// sanity-plugin-webhook-deploy/lib/DeployItem.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _nanoid = require("nanoid");

var _client = _interopRequireDefault(require("part:@sanity/base/client"));

var _styledComponents = _interopRequireDefault(require("styled-components"));

var _ui = require("@sanity/ui");

var _axios = _interopRequireDefault(require("axios"));

var _templateObject, _templateObject2, _templateObject3, _templateObject4, _templateObject5;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var DeployItem = _ref => {
  var hook = _ref.hook;
  var toast = (0, _ui.useToast)();

  var _useState = (0, _react.useState)((0, _nanoid.nanoid)()),
      _useState2 = _slicedToArray(_useState, 2),
      lastDeploy = _useState2[0],
      setLastDeploy = _useState2[1];

  var _useState3 = (0, _react.useState)(false),
      _useState4 = _slicedToArray(_useState3, 2),
      isDeleting = _useState4[0],
      setIsDeleting = _useState4[1];

  var _useState5 = (0, _react.useState)(false),
      _useState6 = _slicedToArray(_useState5, 2),
      isConfirmOpen = _useState6[0],
      setConfirmOpen = _useState6[1];

  var onDelete = /*#__PURE__*/function () {
    var _ref2 = _asyncToGenerator(function* () {
      setIsDeleting(true);

      try {
        yield _client.default.delete(hook._id);
      } catch (error) {
        toast.push({
          status: 'error',
          title: 'An error occured in deleting this webhook.'
        });
      }

      setIsDeleting(false);
      toast.push({
        status: 'success',
        title: "Deleted ".concat(hook.title, "!")
      });
      setConfirmOpen(false);
    });

    return function onDelete() {
      return _ref2.apply(this, arguments);
    };
  }();

  var onDeploy = /*#__PURE__*/function () {
    var _ref3 = _asyncToGenerator(function* () {
      try {
        yield _axios.default.post(hook.url, {});
      } catch (error) {
        toast.push({
          status: 'error',
          title: 'An error occured in deploying this webhook.'
        });
      }

      setLastDeploy((0, _nanoid.nanoid)());
      toast.push({
        status: 'success',
        title: "Deployed ".concat(hook.title, "!")
      });
    });

    return function onDeploy() {
      return _ref3.apply(this, arguments);
    };
  }();

  var confirmDialog = isConfirmOpen && /*#__PURE__*/_react.default.createElement(_ui.Dialog, {
    width: 1,
    header: "Delete ".concat(hook.title),
    onClose: () => setConfirmOpen(false)
  }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
    padding: 4,
    paddingTop: 0
  }, /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginBottom: 4
  }, /*#__PURE__*/_react.default.createElement("p", null, "This deletes your deploy button, but will not delete the environment. This is a permanent action. You'll need to re-add this environment.")), isDeleting ? /*#__PURE__*/_react.default.createElement(_ui.Spinner, null) : /*#__PURE__*/_react.default.createElement(_ui.Button, {
    tone: "critical",
    text: "Permanently Delete",
    onClick: onDelete
  })));

  var CardRow = _styledComponents.default.div(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n\t\tdisplay: flex;\n\t\tflex-direction: column;\n\n\t\t@media (min-width: 1000px) {\n\t\t\tflex-direction: row;\n\t\t\tjustify-content: space-between;\n\t\t\talign-items: center;\n\t\t}\n\t"])));

  var MetaContainer = _styledComponents.default.div(_templateObject2 || (_templateObject2 = _taggedTemplateLiteral(["\n\t\tdisplay: flex;\n\t\tflex-direction: column;\n\t\talign-items: flex-start;\n\t\tmargin-bottom: 20px;\n\n\t\t@media (min-width: 1000px) {\n\t\t\tflex-direction: row;\n\t\t\tmargin-bottom: 0;\n\t\t}\n\t"])));

  var MetaColumn = _styledComponents.default.div(_templateObject3 || (_templateObject3 = _taggedTemplateLiteral(["\n\t\tmargin-bottom: 5px;\n\n\t\t@media (min-width: 1000px) {\n\t\t\tmargin-bottom: 0;\n\t\t\tmargin-right: 10px;\n\t\t}\n\t"])));

  var ActionsContainer = _styledComponents.default.div(_templateObject4 || (_templateObject4 = _taggedTemplateLiteral(["\n\t\tdisplay: flex;\n\t\tflex-direction: row;\n\t\talign-items: flex-start;\n\n\t\t@media (min-width: 1000px) {\n\t\t\talign-items: flex-end;\n\t\t}\n\t"])));

  var Code = _styledComponents.default.div(_templateObject5 || (_templateObject5 = _taggedTemplateLiteral(["\n\t\tfont-family: -apple-system-ui-monospace, 'SF Mono', Menlo, Monaco,\n\t\t\tConsolas, monospace;\n\t\tpadding: 2px;\n\t\tdisplay: block;\n\t\tfont-size: 0.75em;\n\t\toverflow-wrap: break-word;\n\t\tborder-radius: 5px;\n\t\tborder: 1px solid #cdcdcd;\n\t\t@media (max-width: 600px) {\n\t\t\tmax-width: 250px;\n\t\t}\n\t"])));

  return /*#__PURE__*/_react.default.createElement(_ui.Card, {
    padding: 4,
    radius: 2,
    shadow: 1
  }, /*#__PURE__*/_react.default.createElement(CardRow, null, /*#__PURE__*/_react.default.createElement(MetaContainer, null, /*#__PURE__*/_react.default.createElement(MetaColumn, null, /*#__PURE__*/_react.default.createElement("p", {
    style: {
      margin: 0
    }
  }, hook.title || 'Netlify')), /*#__PURE__*/_react.default.createElement(MetaColumn, null, /*#__PURE__*/_react.default.createElement(Code, {
    size: 1
  }, hook.url)), /*#__PURE__*/_react.default.createElement("img", {
    key: lastDeploy,
    src: hook.statusBadgeUrl,
    style: {
      display: 'block'
    }
  })), /*#__PURE__*/_react.default.createElement(ActionsContainer, null, /*#__PURE__*/_react.default.createElement(_ui.Box, null, /*#__PURE__*/_react.default.createElement(_ui.Button, {
    tone: "critical",
    text: "Delete",
    onClick: () => setConfirmOpen(true)
  })), /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginLeft: 2
  }, /*#__PURE__*/_react.default.createElement(_ui.Button, {
    tone: "primary",
    text: "Start Deploy",
    onClick: onDeploy
  })))), confirmDialog);
};

var _default = DeployItem;
exports.default = _default;
